### DISCUSSION TITLE
How do you get the duration of a GIF?

### DISCUSSION THREAD

**Discussion by natanfudge** [coil-kt/coil#1264] [Q&A] [answered]
Hey, I'm thinking of displaying a duration indicator at the bottom of the screen when displaying a GIF. However, I couldn't find any way to actually get this information from the Coil API. I understand that you're supposed to sum the delays of all frames, but how do you actually get the frame delays in the Coil context?

**Comment by colinrtwhite** [comment]
I don't think it's possible to get the GIF duration at the moment. To add this feature [ImageDecoder](https://developer.android.com/reference/android/graphics/ImageDecoder) would need to expose that info as well.

**Reply by natanfudge** [reply]
@colinrtwhite 
Yes, it seems that once it goes through `ImageDecoder` that information is lost. Maybe there's a way to gain access to the `ByteArray`/`InputStream` of the gif as it was received from the network, and parse the duration information myself, using a separate API? It would be slower, but it would get the job done (and this is a very important feature for me!)

**Reply by colinrtwhite** [reply]
Your best bet would be to write a custom `Decoder.Factory` that delegates to `ImageDecoderDecoder.Factory`/`GifDecoder.Factory` then reads the duration from the `ImageSource` if necessary.

**Comment by natanfudge** [comment, marked as answer]
Using @colinrtwhite advice and the [Metadata Extractor Library](https://github.com/drewnoakes/metadata-extractor), I've implemented this feature. For those looking for the same thing, I've posted a full gist of the code: https://gist.github.com/natanfudge/709c6a86097da4162fa47074e632317b. 

Note it does use some reflection to access internal Coil classes, it would be nice if you could find an alternative!
